File: spritesheet.py
from PIL import Image

class Spritesheet:

    def __init__(self, src, frame_width, frame_height, frame_count):
        self._loaded_correctly = False
        
        try:
            self._full_spriteheet = Image.open(src)

        except Exception as e:
            print("Failed to load image! Reason:")
            print(e)

            return self

        self._frame_width = frame_width
        self._frame_height = frame_height
        self._frame_count = frame_count

        num_frames_h = int(self._full_spriteheet.width / frame_width)
        num_frames_v = int(self._full_spriteheet.height / frame_height)
        
        self._frames = []
        break_outer_loop = False

        for i in range(num_frames_v):
            for j in range(num_frames_h):
                if (i * num_frames_h + j) >= frame_count:
                    break_outer_loop = True
                    break
                
                crop_region = (j * frame_width, i * frame_height,
                                (j + 1) * frame_width, (i + 1) * frame_height)
                self._frames.append(self._full_spriteheet.crop(crop_region))

            if break_outer_loop:
                break

        self._loaded_correctly = True

       

    def get_frame(self, frame_number : float) -> Image:
        """
        
        """
        
        while frame_number >= self._frame_count:
            frame_number -= self._frame_count

        while frame_number < 0:
            frame_number += self._frame_count

        return self._frames[int(frame_number)]

        
        

File: static/gui/static/js/components/Expression/Spritesheet.js
export class SpritesheetJS{
    constructor(src, frameWidth, frameHeight, frameCount) {
        this.loadedCorrectly = false;
        this.frameWidth = frameWidth;
        this.frameHeight = frameHeight;
        this.frameCount = frameCount;
        this.imageSrc = src;

        console.log(`Image Src: ${src}`);

        this.loaded = new Promise((resolve, reject) => {
            this.fullSpritesheet = new Image();
            this.fullSpritesheet.src = src;
            this.fullSpritesheet.onload = () => {
                this.loadedCorrectly = true;
                resolve();
            };
            this.fullSpritesheet.onerror = () => {
                this.loadedCorrectly = false;
                reject(new Error('Failed to load spritesheet'));
              };
        });
    }

    /**
     * Draw the specified frame on the specified canvas.
     * @param {number} frameNumber - The number of the frame to draw.
     * @param {number} canvas - The canvas element on which to draw the frame.
     * @returns {boolean} Whether the frame was successfully drawn on the canvas.
     */
    drawFrame(frameNumber, canvas) {
        if (!this.loadedCorrectly || !canvas)
            return false;
    
        while (frameNumber >= this.frameCount) {
            frameNumber -= this.frameCount;
        }
        while (frameNumber < 0) {
            frameNumber += this.frameCount;
        }

        const frameIndex = Math.floor(frameNumber);
        const numFramesH = Math.floor(this.fullSpritesheet.width / this.frameWidth);
        const row = Math.floor(frameIndex / numFramesH);
        const col = frameIndex % numFramesH;
        const sx = col * this.frameWidth;
        const sy = row * this.frameHeight;
        const context = canvas.getContext('2d');
        context.drawImage(this.fullSpritesheet, sx, sy, this.frameWidth, this.frameHeight, 0, 0, canvas.width, canvas.height);
        return true;
    }
}